pkg/kube/object_patch/operation.go

```diff
@@ -0,0 +1,253 @@
+			WithSubresource(spec.Subresource),
+			InBackground())
+	case DeleteNonCascading:
+		return NewDeleteOperation(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			NonCascading())
+	case JQPatch:
+		return NewFilterPatchOperation(func(u *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+			return applyJQPatch(spec.JQFilter, u)
+		},
+			spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		)
+	case MergePatch:
+		return NewMergePatchOperation(spec.MergePatch,
+			spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		)
+		//mergePatch, err := json.Marshal(spec.MergePatch)
+		//if err != nil {
+		//	applyErrors = multierror.Append(applyErrors, err)
+		//	continue
+		//}
+		//operationError = o.Patch(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+		//	WithSubresource(spec.Subresource),
+		//	UseMergePatch(mergePatch),
+		//	WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		//)
+	case JSONPatch:
+		return NewJSONPatchOperation(spec.JSONPatch,
+			spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		)
+		//jsonPatch, err := json.Marshal(spec.JSONPatch)
+	}
+
+	// Should not be reached!
+	return nil
+}
+
+func NewCreateOperation(obj interface{}, options ...OperationOption) *Operation {
+	op := &Operation{
+		object: obj,
+	}
+	op.applyOptions(options...)
+	return op
+}
+
+func NewDeleteOperation(apiVersion, kind, namespace, name string, options ...OperationOption) *Operation {
+	op := &Operation{
+		apiVersion:          apiVersion,
+		kind:                kind,
+		namespace:           namespace,
+		name:                name,
+		deletionPropagation: metav1.DeletePropagationForeground,
+	}
+	op.applyOptions(options...)
+	return op
+}
+
+func NewMergePatchOperation(mergePatch interface{}, apiVersion, kind, namespace, name string, options ...OperationOption) *Operation {
+	op := &Operation{
+		apiVersion: apiVersion,
+		kind:       kind,
+		namespace:  namespace,
+		name:       name,
+		patch:      mergePatch,
+		patchType:  types.MergePatchType,
+	}
+	op.applyOptions(options...)
+	return op
+}
+
+func NewJSONPatchOperation(jsonPatch interface{}, apiVersion, kind, namespace, name string, options ...OperationOption) *Operation {
+	op := &Operation{
+		apiVersion: apiVersion,
+		kind:       kind,
+		namespace:  namespace,
+		name:       name,
+		patch:      jsonPatch,
+		patchType:  types.JSONPatchType,
+	}
+	op.applyOptions(options...)
+	return op
+}
+
+func NewFilterPatchOperation(filter func(*unstructured.Unstructured) (*unstructured.Unstructured, error), apiVersion, kind, namespace, name string, options ...OperationOption) *Operation {
```

Comment:
There is no NewJQPatchOperation. JQPatch is now available for shell hooks and not available for Go hooks.